ZuluContent/Skills/Meditation.cs

```diff
@@ -1,80 +1,218 @@
+        {
+            SkillInfo.Table[46].Callback = OnUse;
+        }
+        
+        public static TimeSpan OnUse(Mobile m)
+        {
+            m.RevealingAction();
+
+            if (m.Mana >= m.ManaMax)
+            {
+                m.SendLocalizedMessage(501846); // You are at peace.
+                return DefaultDelay;
+            }
+
+            if (m.Poisoned)
+            {
+                m.SendAsciiMessage("You can't meditate while poisoned.");
+                return DefaultDelay;
+            }
+
+            if (m.Warmode)
+            {
+                m.SendAsciiMessage("You can't meditate in war mode.");
+                return DefaultDelay;
+            }
+
+            if (!CheckValidHands(m))
+            {
+                m.SendLocalizedMessage(502626); // Your hands must be free to cast spells or meditate.
+                return DefaultDelay;
+            }
+            
+            if (GetMagicEfficiencyModifier(m) <= 0)
+            {
+                m.SendAsciiMessage("Regenerative forces cannot penetrate your armor.");
+                return DefaultDelay;
+            }
+
+            if ((m as IShilCheckSkill)?.CheckSkill(SkillName.Meditation, -1,
+                SkillCheck.Configs[SkillName.Meditation].DefaultPoints) == true)
+            {
+                m.SendLocalizedMessage(501851); // You enter a meditative trance.
+                m.Meditating = true;
+
+                m.PlaySound(0xFA);
+                
+                var regenBase = (int) (m.Skills[SkillName.Meditation].Value / 25 + m.Int / 35.0);
+                var interval = 5.0;
+
+                m.FireHook(h => h.OnMeditation(m, ref regenBase, ref interval));
+                
+                new InternalTimer( m, regenBase, TimeSpan.FromSeconds(interval)).Start();
+                return TimeSpan.FromSeconds(10.0);
+            }
+
+            
+            m.SendAsciiMessage("You cannot focus your concentration.");
+            return DefaultDelay;
+        }
+
+        public static bool CheckValidHands(Mobile m)
+        {
+            var items = new[] {m.FindItemOnLayer(Layer.OneHanded), m.FindItemOnLayer(Layer.TwoHanded)};
+            
+            return items.All(item =>
+            {
+                switch (item)
+                {
+                    case null:
+                    case Spellbook:
+                    case Runebook:
+                    case BaseStaff:
+                        return true;
+                    case IEnchanted enchanted:
+                    {
+                        var magical = enchanted.Enchantments.Get((MagicalWeapon e) => e.Value);
+                        return magical == MagicalWeaponType.Mystical || magical == MagicalWeaponType.Stygian;
+                    }
+                    default:
+                        return false;
+                }
+            });
+        }
+
+        public static double GetMagicEfficiencyModifier(Mobile from)
```

Comment:
Mmm just wondering if this should belong here, it will be needed in many other places (eg. calcs for how many summoned creatures you can control, spelldamage etc.) and would feel weird to import it from the Meditation class